File: hello-typescript/src/server/network/AudioWebSocketServer.ts
/**
 * WebSocket server for audio streaming.
 * Handles client connections and message routing.
 * Matches C++ WebSocketServer and Java AudioWebSocketServer functionality.
 */

import * as WebSocket from "ws";
import { StreamManager } from "../memory/StreamManager";
import { MemoryPoolManager } from "../memory/MemoryPoolManager";
import { WebSocketMessageHandler } from "../handler/WebSocketMessageHandler";

export class AudioWebSocketServer {
  private host: string;
  private port: number;
  private path: string;
  private server: WebSocket.Server | null;
  private streamManager: StreamManager;
  private memoryPool: MemoryPoolManager;
  private connectionHandlers: Map<WebSocket, WebSocketMessageHandler>;

  constructor(
    host: string,
    port: number,
    path: string,
    streamManager: StreamManager,
  ) {
    this.host = host;
    this.port = port;
    this.path = path;
    this.server = null;
    this.streamManager = streamManager;
    this.memoryPool = MemoryPoolManager.getInstance();
    this.connectionHandlers = new Map();

    console.log(`AudioWebSocketServer initialized on ${host}:${port}${path}`);
  }

  start(): void {
    this.server = new WebSocket.Server({
      host: this.host,
      port: this.port,
      path: this.path,
      maxPayload: 100 * 1024 * 1024, // 100MB max message size
    });

    this.server.on("connection", (ws: WebSocket) => {
      this.handleConnection(ws);
    });

    this.server.on("error", (error: Error) => {
      console.error("WebSocket server error:", error);
    });

    console.log(
      `WebSocket server started on ws://${this.host}:${this.port}${this.path}`,
    );
  }

  stop(): void {
    if (this.server) {
      this.server.close(() => {
        console.log("WebSocket server stopped");
      });
    }
  }

  private handleConnection(ws: WebSocket): void {
    const clientAddr = `${(ws as any)._socket?.remoteAddress}:${(ws as any)._socket?.remotePort}`;
    console.log(`Client connected: ${clientAddr}`);

    // Create a handler for this connection
    const handler = new WebSocketMessageHandler(this.streamManager);
    this.connectionHandlers.set(ws, handler);

    ws.on("message", (message: WebSocket.Data) => {
      handler.handleMessage(ws, message);
    });

    ws.on("close", () => {
      console.log(`Client disconnected: ${clientAddr}`);
      this.connectionHandlers.delete(ws);
    });

    ws.on("error", (error: Error) => {
      console.error(`Client error ${clientAddr}:`, error);
      this.connectionHandlers.delete(ws);
    });
  }
}

File: hello-python/src/audio_server/network/audio_websocket_server.py
"""
WebSocket server for audio streaming.
Handles client connections and message routing.
Matches C++ WebSocketServer and Java AudioWebSocketServer functionality.
"""

import asyncio
from typing import Set
import websockets
from websockets.asyncio.server import serve, ServerConnection
from loguru import logger

from ..memory.stream_manager import StreamManager
from ..memory.memory_pool_manager import MemoryPoolManager
from ..handler.websocket_message_handler import WebSocketMessageHandler


class AudioWebSocketServer:
    """
    WebSocket server for handling audio stream uploads and downloads.
    Manages client connections and routes messages to handler.
    """

    def __init__(self, host: str = "0.0.0.0", port: int = 8080, path: str = "/audio"):
        """
        Initialize WebSocket server.

        Args:
            host: Host address to bind to
            port: Port number to listen on
            path: WebSocket endpoint path
        """
        self.host = host
        self.port = port
        self.path = path
        self.clients: Set[ServerConnection] = set()
        self.stream_manager = StreamManager.get_instance()
        self.memory_pool = MemoryPoolManager.get_instance()
        self.message_handler = WebSocketMessageHandler(self.stream_manager)
        self.server = None

        logger.info(f"AudioWebSocketServer initialized on {host}:{port}{path}")

    async def start(self):
        """Start the WebSocket server"""
        self.server = await serve(
            self.handle_client,
            self.host,
            self.port,
            max_size=100 * 1024 * 1024,  # 100MB max message size
            ping_interval=20,
            ping_timeout=20,
        )
        logger.info(
            f"WebSocket server started on ws://{self.host}:{self.port}{self.path}"
        )

    async def stop(self):
        """Stop the WebSocket server"""
        if self.server:
            self.server.close()
            await self.server.wait_closed()
            logger.info("WebSocket server stopped")

    async def handle_client(self, websocket: ServerConnection):
        """
        Handle a client connection.

        Args:
            websocket: WebSocket connection
        """
        # Register client
        self.clients.add(websocket)
        client_addr = f"{websocket.remote_address[0]}:{websocket.remote_address[1]}"
        logger.info(f"Client connected: {client_addr}")

        try:
            async for message in websocket:
                await self.message_handler.handle_message(websocket, message)
        except Exception as e:
            logger.error(f"Error handling client {client_addr}: {e}")
        finally:
            # Unregister client
            self.clients.discard(websocket)
            logger.info(f"Client disconnected: {client_addr}")


async def main():
    """Main entry point for running the server standalone"""
    server = AudioWebSocketServer(host="0.0.0.0", port=8080, path="/audio")
    await server.start()

    # Keep server running
    try:
        await asyncio.Future()  # Run forever
    except KeyboardInterrupt:
        logger.info("Shutting down server...")
        await server.stop()


if __name__ == "__main__":
    asyncio.run(main())
